Merge all features related to C# 7 ref features (#795)
dotnet/csharpstandard: standard/expressions.md

- For events, the argument list consists of the expression specified as the right operand of the `+=` or `-=` operator.
- For user-defined operators, the argument list consists of the single operand of the unary operator or the two operands of the binary operator.

> The arguments of properties ([§15.7](classes.md#157-properties)), events ([§15.8](classes.md#158-events)), and user-defined operators ([§15.10](classes.md#1510-operators)) are always passed as value parameters ([§15.6.2.2](classes.md#15622-value-parameters)). The arguments of indexers ([§15.9](classes.md#159-indexers)) are always passed as value parameters ([§15.6.2.2](classes.md#15622-value-parameters)) or parameter arrays ([§15.6.2.5](classes.md#15625-parameter-arrays)). Input, output, and reference parameters are not supported for these categories of function members.

Reviewer: >Input, [](http://example.com/codeflow?start=478&length=6)

Aren't `in` parameters allowed for indexers?

Author: Yes, input parameters are also for user defined operators.

> The arguments of properties ([§15.7](classes.md#157-properties)) and events ([§15.8](classes.md#158-events)) are always passed as value parameters ([§15.6.2.2](classes.md#15622-value-parameters)). The arguments of user-defined operators ([§15.10](classes.md#1510-operators)) are always passed as value parameters ([§15.6.2.2](classes.md#15622-value-parameters)) or input parameters (§input-parameters-new-clause). The arguments of indexers ([§15.9](classes.md#159-indexers)) are always passed as value parameters ([§15.6.2.2](classes.md#15622-value-parameters)), input parameters (§input-parameters-new-clause), or parameter arrays ([§15.6.2.5](classes.md#15625-parameter-arrays)). Output, and reference parameters are not supported for these categories of function members.

Reviewer 2: Not sure that the change from "Reference and output parameters" to "Output, and reference parameters" is beneficial. Was this perhaps a change to "Input, output and reference parameters" followed by a realization that we didn't actually want to include output parameters?

Author: It was to be consistent. The order for output, reference had been different in different locations. When Rex started these edits, I think he swept through everywhere and used "input, output, and reference" (analogy with read, write, and read-write). Where input is missing, it becomes "output and reference."

Not resolving for discussion.

Reviewer 2: ```suggestion
The arguments of properties ([§15.7](classes.md#157-properties)) and events ([§15.8](classes.md#158-events)) are always passed as value parameters ([§15.6.2.2](classes.md#15622-value-parameters)). The arguments of user-defined operators ([§15.10](classes.md#1510-operators)) are always passed as value parameters ([§15.6.2.2](classes.md#15622-value-parameters)) or input parameters (§input-parameters-new-clause). The arguments of indexers ([§15.9](classes.md#159-indexers)) are always passed as value parameters ([§15.6.2.2](classes.md#15622-value-parameters)), input parameters (§input-parameters-new-clause), or parameter arrays ([§15.6.2.5](classes.md#15625-parameter-arrays)). Output and reference parameters are not supported for these categories of function members.
```

- The keyword `ref` followed by a *variable_reference* ([§9.5](variables.md#95-variable-references)), indicating that the argument is passed as a reference parameter ([§15.6.2.3](classes.md#15623-reference-parameters)). A variable shall be definitely assigned ([§9.4](variables.md#94-definite-assignment)) before it can be passed as a reference parameter.
- The keyword `out` followed by a *variable_reference* ([§9.5](variables.md#95-variable-references)), indicating that the argument is passed as an output parameter ([§15.6.2.4](classes.md#15624-output-parameters)). A variable is considered definitely assigned ([§9.4](variables.md#94-definite-assignment)) following a function member invocation in which the variable is passed as an output parameter.

The form determines the ***parameter-passing mode*** of the argument: *value*, *input*, *reference*, or *output*, respectively. However, as mentioned above, an argument with value passing mode, might be transformed into one with input passing mode.

> Passing a volatile field ([§15.5.4](classes.md#1554-volatile-fields)) as an input, output, or reference parameter causes a warning, since the field may not be treated as volatile by the invoked method.

Reviewer 3: Should this be moved to the unsafe chapter?

Reviewer 2: We'll fix this post-C# 7. Bill has created #800.

During the run-time processing of a function member invocation ([§12.6.6](expressions.md#1266-function-member-invocation)), the expressions or variable references of an argument list are evaluated in order, from left to right, as follows:

> - For a value parameter, if the parameter’s passing mode is value

Reviewer: >For a value parameter, if the parameter’s passing mode is value [](http://example.com/codeflow?start=2&length=63)

Does "value parameter" include parameters declared with an `in` modifier? Given that the next main bullet describes "input, output, or reference" parameters, it seems "value parameter" is a parameter without a modifier, and in that case, isn't the "parameter's passing mode" always "value"?

Author: That first phrase should be "For a value argument, ..."

  - the argument expression is evaluated and an implicit conversion ([§10.2](conversions.md#102-implicit-conversions)) to the corresponding parameter type is performed. The resulting value becomes the initial value of the value parameter in the function member invocation.
  - otherwise, the parameter’s passing mode is input, in which case, a storage location is created with the same type as that of the corresponding parameter. The argument expression is evaluated and an implicit conversion ([§10.2](conversions.md#102-implicit-conversions)) to the corresponding parameter type is performed. The resulting value is stored within that storage location. That storage location is represented by the input parameter in the function member invocation.

     > *Example*: Given the following declarations and method calls:
     >
     > ```csharp
     > public static void M1(in int p1) { … }
     > int i = 10;

>      > M1(i);         // transformed to a temporary input argument

Reviewer: >M1(i);         // transformed to a temporary input argument [](http://example.com/codeflow?start=7&length=59)

Is there a requirement that the compiler generate a temporary in this case?

Author: @cston That's been my understanding: If the `in` modifier isn't specified on the argument, a temporary must be created.

Reviewer 2: It looks like that might actually not always be the case. Consider this:

```csharp
class Program
{
    static void Main()
    {
        int i = 5;
        M(i, () => i++);
    }

    static void M(in int x, Action action)
    {
        Console.WriteLine(x);
        action();
        Console.WriteLine(x);
    }
}
```

On my machine, that *unexpectedly to me* produces output of 5 then 6. I'd *expected* the value of `i` to be passed to the method, displaying 5 twice. Changing the invocation to `M(i + 0, () => i++);` changes the output to 5 twice.

Personally I'd suggest that's a Roslyn bug :)

Author: This is one for @MadsTorgersen , but I'm tempted to agree with ~~out~~ you.

Reviewer 2: Not sure what you mean by "tempted to agree with out" - I *suspect* "out" was a typo for "you", but I don't want to assume.

Reviewer 2: @ericlippert and @gafter suggest that the behavior is intentional, which means we need to fix the text (which says there's *always* a new storage location).

@BillWagner will fix this, both in the text and example, to special case the "argument is a variable of a type which has an identity conversion to the parameter type".

Reviewer 3: From https://github.com/dotnet/csharplang/blob/main/meetings/2017/LDM-2017-09-27.md

> At the call site, in is optional. If it is used, the argument has to an lvalue that can be
> passed directly by ref. If not, then the compiler does its best: copy only if necessary.

     > M1(i + 5);     // transformed to a temporary input argument
     > ```
     >
     > In each method call, an unnamed `int` variable is created, initialized with the argument’s value, and then passed as an input argument. See [§12.6.4.2](expressions.md#12642-applicable-function-member) and §better-argument-passing-mode-new-clause.
     >
     > *end example*

> - For an input, output, or reference parameter, the variable reference is evaluated and the resulting storage location becomes the storage location represented by the parameter in the function member invocation. If the variable reference given as an output, or reference is an array element of a *reference_type*, a run-time check is performed to ensure that the element type of the array is identical to the type of the parameter. If this check fails, a `System.ArrayTypeMismatchException` is thrown.

Reviewer 2: This changes from talking about a value *argument* to an input/output/reference *parameter*. I think this should be "argument", given that it really is talking about the passing mode of the argument.

Author: I think it's correct, as this is defining how an argument is passed as the parameter.

Leaving unresolved for discussion.

Reviewer 3: "For an input, output, or reference parameter" should be "For an input, output, or reference argument".

Should we mention here that, for `in` and `ref` modes, the variable must be definitely assigned?

>
> *end example*

> The array co-variance rules ([§17.6](arrays.md#176-array-covariance)) permit a value of an array type `A[]` to be a reference to an instance of an array type `B[]`, provided an implicit reference conversion exists from `B` to `A`. Because of these rules, when an array element of a *reference_type* is passed as an input, output, or reference argument, a run-time check is required to ensure that the actual element type of the array is *identical* to that of the parameter.

Reviewer 3: Perhaps "is required" should be "occurs".  Should we mention what the check is and what happens when it fails?  Should we give an example?

Reviewer 4: "Occurs" doesn't imply "required" to me, and the Standard is requiring (often "must" in the Standard)

Maybe something along the lines of "a run-time check (is required OR must be performed) to check ... and if this fails an error produced"

Reviewer 2: Punting until post-C# 7 - it's not a new problem, if it's a problem at all.
Bill has raised this in #801.

> The array co-variance rules ([§17.6](arrays.md#176-array-covariance)) permit a value of an array type `A[]` to be a reference to an instance of an array type `B[]`, provided an implicit reference conversion exists from `B` to `A`. Because of these rules, when an array element of a *reference_type* is passed as an input, output, or reference argument, a run-time check is required to ensure that the actual element type of the array is *identical* to that of the parameter.

Reviewer 5: The runtime check should be skipped for an input parameter here too, like in [the ArrayTypeMismatchException paragraph above](https://github.com/dotnet/csharpstandard/blob/3c50a813cad5328eb6709aadffb34873696dccb8/standard/expressions.md?plain=1#L618).  <https://github.com/dotnet/csharpstandard/issues/264#issuecomment-874738023>

```suggestion
The array co-variance rules ([§17.6](arrays.md#176-array-covariance)) permit a value of an array type `A[]` to be a reference to an instance of an array type `B[]`, provided an implicit reference conversion exists from `B` to `A`. Because of these rules, when an array element of a *reference_type* is passed as an output or reference argument, a run-time check is required to ensure that the actual element type of the array is *identical* to that of the parameter.
```

Reviewer 5: *identical* here means any pair of types that has an identity conversion, right?  So `object` and `dynamic` are *identical*, and the run-time check would not be able to distinguish them anyway.  Tuple types have identity conversions too, but they are value types so array covariance does not apply.

> The array co-variance rules ([§17.6](arrays.md#176-array-covariance)) permit a value of an array type `A[]` to be a reference to an instance of an array type `B[]`, provided an implicit reference conversion exists from `B` to `A`. Because of these rules, when an array element of a *reference_type* is passed as an input, output, or reference parameter, a run-time check is required to ensure that the actual element type of the array is *identical* to that of the parameter.

Reviewer 2: "Passed as [...] parameter" doesn't feel right to me. Can we change this one to "argument" as well?

- If `Eᵢ` is an anonymous function, an *explicit parameter type inference* ([§12.6.3.8](expressions.md#12638-explicit-parameter-type-inferences)) is made *from* `Eᵢ` *to* `Tᵢ`
- Otherwise, if `Eᵢ` has a type `U` and `xᵢ` is a value parameter ([§15.6.2.2](classes.md#15622-value-parameters)) then a *lower-bound inference* ([§12.6.3.10](expressions.md#126310-lower-bound-inferences)) is made *from* `U` *to* `Tᵢ`.
- Otherwise, if `Eᵢ` has a type `U` and `xᵢ` is a reference parameter ([§15.6.2.3](classes.md#15623-reference-parameters)), or output parameter ([§15.6.2.4](classes.md#15624-output-parameters)) then an *exact inference* ([§12.6.3.9](expressions.md#12639-exact-inferences)) is made *from* `U` *to* `Tᵢ`.

> - Otherwise, if `Eᵢ` has a type `U` and `xᵢ` is an input parameter (§method-input-parameters-new-clause) then a *lower bound inference* ([§12.6.3.10](expressions.md#126310-lower-bound-inferences)) is made *from* `U` *to* `Tᵢ`. A new storage location is created following the rules of §12.6.2.2 on corresponding parameters.

Reviewer: >A new storage location is created following the rules of §12.6.2.2 on corresponding parameters. [](http://example.com/codeflow?start=227&length=95)

Should this be described as part of the invocation expression rather than type inference?

Author: I think it's needed here. The purpose is to allow implicit conversions to an `in` parameter when the argument doesn't include the `in` modifier. For example:

```csharp
public void M(in long p) {}

// calling:

long num = 25;
int i = 19;

M(in i); // Not allowed. Implicit conversion, not an identity conversion.
M (i); // OK, implicit conversion and the creation of temporary.
M(in num); // OK
M(num); // OK.

Reviewer: Does the creation of a storage location affect type inference though?

Author: Leaving unresolved for the committee to weigh in. I agree with @cston , that sentence can be removed here.

Reviewer 3: Missing description of exact inference from an input `in` argument to an `in` parameter, which requires exact inference.

Reviewer 2: > Missing description of exact inference from an input in argument to an in parameter, which requires exact inference.

@BillWagner has fixed this (in a new previous bullet)

@BillWagner will commit a suggestion of removing the "storage location" sentence in this bullet.

> - Otherwise, if `Eᵢ` has a type `U` and `xᵢ` is an input parameter (§method-input-parameters-new-clause) then a *lower bound inference* ([§12.6.3.10](expressions.md#126310-lower-bound-inferences)) is made *from* `U` *to* `Tᵢ`. A new storage location is created following the rules of §12.6.2.2 on corresponding parameters.

Reviewer 3: This doesn't correctly handle an `in` argument to an input parameter.  That requires exact inference.

Author: Added a bullet item for `in` parameter and `in` argument.

- If `Eᵢ` is an anonymous function, an *explicit parameter type inference* ([§12.6.3.8](expressions.md#12638-explicit-parameter-type-inferences)) is made *from* `Eᵢ` *to* `Tᵢ`
- Otherwise, if `Eᵢ` has a type `U` and `xᵢ` is a value parameter ([§15.6.2.2](classes.md#15622-value-parameters)) then a *lower-bound inference* ([§12.6.3.10](expressions.md#126310-lower-bound-inferences)) is made *from* `U` *to* `Tᵢ`.
- Otherwise, if `Eᵢ` has a type `U` and `xᵢ` is a reference parameter ([§15.6.2.3](classes.md#15623-reference-parameters)), or output parameter ([§15.6.2.4](classes.md#15624-output-parameters)) then an *exact inference* ([§12.6.3.9](expressions.md#12639-exact-inferences)) is made *from* `U` *to* `Tᵢ`.
- Otherwise, if `Eᵢ` has a type `U` and `xᵢ` is an input parameter (§method-input-parameters-new-clause) and `Ei` is an input argument, then an *exact inference* ([§12.6.3.9](expressions.md#12639-exact-inferences)) is made *from* `U` *to* `Tᵢ`.

> - Otherwise, if `Eᵢ` has a type `U` and `xᵢ` is an input parameter (§method-input-parameters-new-clause) then a *lower bound inference* ([§12.6.3.10](expressions.md#126310-lower-bound-inferences)) is made *from* `U` *to* `Tᵢ`. A new storage location is created following the rules of §12.6.2.2 on corresponding parameters.

Author: ```suggestion
- Otherwise, if `Eᵢ` has a type `U` and `xᵢ` is an input parameter (§method-input-parameters-new-clause) then a *lower bound inference* ([§12.6.3.10](expressions.md#126310-lower-bound-inferences)) is made *from* `U` *to* `Tᵢ`.
```

- For each argument in `A`, the parameter-passing mode of the argument is identical to the parameter-passing mode of the corresponding parameter, and
  - for a value parameter or a parameter array, an implicit conversion ([§10.2](conversions.md#102-implicit-conversions)) exists from the argument expression to the type of the corresponding parameter, or
  - for a `ref` or `out` parameter, there is an identity conversion between the type of the argument expression (if any) and the type of the corresponding parameter
  - for an `in` parameter when the corresponding argument has the `in` modifier, there is an identity conversion between the type of the argument expression (if any) and the type of the corresponding parameter

>   - for an `in` parameter when the corresponding argument omits the `in` modifier, an implicit conversion ([§10.2](conversions.md#102-implicit-conversions)), excluding dynamic implicit conversions (§10.2.10) exists from the argument expressions to the type of the corresponding parameter.

Reviewer: >argument expressions [](http://example.com/codeflow?start=224&length=20)

Should this be "argument expression" singular?

- The expanded form is constructed by replacing the parameter array in the function member declaration with zero or more value parameters of the element type of the parameter array such that the number of arguments in the argument list `A` matches the total number of parameters. If `A` has fewer arguments than the number of fixed parameters in the function member declaration, the expanded form of the function member cannot be constructed and is thus not applicable.
- Otherwise, the expanded form is applicable if for each argument in `A`, one of the following is true:
  - the parameter-passing mode of the argument is identical to the parameter-passing mode of the corresponding parameter, and
    - for a fixed value parameter or a value parameter created by the expansion, an implicit conversion ([§10.2](conversions.md#102-implicit-conversions)) exists from the argument expression to the type of the corresponding parameter, or

>     - for `in`, `out`, or `ref` parameter, the type of the argument expression is identical to the type of the corresponding parameter.

Reviewer 2: ```suggestion
    - for an `in`, `out`, or `ref` parameter, the type of the argument expression is identical to the type of the corresponding parameter.
```

>   - the parameter-passing mode of the argument is value, and the parameter-passing mode of the corresponding parameter is input, and an implicit conversion ([§10.2](conversions.md#102-implicit-conversions)) excluding dynamic implicit conversions (§10.2.10) exists from the argument expression to the type of the corresponding parameter

Reviewer 2: That would suggest this code shouldn't compile - but it does:

```csharp
class Program
{
    static void Main()
    {
        dynamic d = "";
        M(d);
    }

    static void M(in int x)
    {
    }
}
```

Have I misunderstood this?

Author: This is the one area I'm really still fuzzy on how to explain.

Yes, it does compile, but it will throw a `RuntimeBinderException` at runtime. See [Sharplab.io](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7AHwAEAmABgFgAoUgRmutIAIjaB2agb2qd5dqQsUTALIBDAJZ4AFAEoefblT4qmAEwCeeMQFsJYdUwC8TAESmA3AtWjpa2VeV8Avgye9WgosJHSpTKQwmBHl3JiUVVyoooA)

Where I could use help is that there are examples where these conversions don't compile, and others where they compile but always throw exceptions. I'm not sure what conditions cause which.

Hope @gafter can shed light on some of this.

Reviewer 2: Resolving this, but Bill has raised #802 to make sure we fix before C# 7 ships.

- Otherwise if one member is a non-lifted operator and the other is a lifted operator, the non-lifted one is better.
- If neither function member was found to be better, and all parameters of `Mᵥ` have a corresponding argument whereas default arguments need to be substituted for at least one optional parameter in `Mₓ`, then `Mᵥ` is better than `Mₓ`. Otherwise, no function member is better.

#### §better-argument-passing-mode-new-clause Better argument-passing mode

It is permitted to have corresponding parameters in two overloaded methods differ only by parameter-passing mode provided one of the two parameters has value-passing mode, as follows:

```csharp
public static void M1(int p1) { … }
public static void M1(in int p1) { … }
```

> Given `int i = 10;`, according to [§12.6.4.2](expressions.md#12642-applicable-function-member), the calls `M1(i)` and `M1(i + 5)` result in both overloads being applicable. In such cases, the method with the parameter-passing mode of value is the ***better parameter-passing mode choice***.

Reviewer 2: The title of the subsection is better argument-passing mode - should it actually be "better parameter-passing mode"? (I suspect it should.)

Author: Leaving unresolved for committee discussion.

Reviewer 2: @BillWagner: We agreed to change the title to "better parameter-passing mode". Let's do that in this PR.

- Otherwise if one member is a non-lifted operator and the other is a lifted operator, the non-lifted one is better.

> - If neither function member was found to be better, and all parameters of `Mᵥ` have a corresponding argument whereas default arguments need to be substituted for at least one optional parameter in `Mₓ`, then `Mᵥ` is better than `Mₓ`. Otherwise, no function member is better.

Reviewer 2: @BillWagner: While editing these rules, can we put "Otherwise, no function member is better" in its own bullet. (Only if this set of rules does get edited.)

Author: New rule added, and final "otherwise" in its own bullet.

> #### §better-argument-passing-mode-new-clause Better argument-passing mode

Reviewer 2: Comment from @gafter: We don't seem to use this anywhere. Presumably it should be somewhere in the tie-break rules?

Reviewer 2: @BillWagner: This should be resolved before this PR goes in. We need to work out where this should come in the rules.

Author: Proposal is dotnet/csharplang#945.  The tiebreaker rule for `in` vs. ByVal goes last in the list.

See also [LDM notes](https://github.com/dotnet/csharplang/blob/main/meetings/2017/LDM-2017-12-04.md#in-vs-value-parameter-ambiguity)

    ;
```

> The *argument_list* of an *element_access* is not allowed to contain `in`, `out`, or `ref` arguments.

Reviewer: >`in` [](http://example.com/codeflow?start=69&length=4)

Aren't `in` arguments allowed?

Reviewer 5: It seems they are; this builds and runs OK with LangVersion 7.3.

```csharp
public class C {
    public char this[in int index] => '_';
    
    public static void Main()
    {
        C c = new C();
        int index = 0;
        System.Console.WriteLine(c[in index]);
    }
}
```

- When `this` is used in a *primary_expression* within an instance method or instance accessor of a class, it is classified as a value. The type of the value is the instance type ([§15.3.2](classes.md#1532-the-instance-type)) of the class within which the usage occurs, and the value is a reference to the object for which the method or accessor was invoked.
- When `this` is used in a *primary_expression* within an instance constructor of a struct, it is classified as a variable. The type of the variable is the instance type ([§15.3.2](classes.md#1532-the-instance-type)) of the struct within which the usage occurs, and the variable represents the struct being constructed.
  - If the constructor declaration has no constructor initializer, the `this` variable behaves exactly the same as an `out` parameter of the struct type. In particular, this means that the variable shall be definitely assigned in every execution path of the instance constructor.

>   - Otherwise, the `this` variable behaves exactly the same as a `ref` parameter of the struct type. In particular, this means that the variable is considered initially assigned, and can't be reassigned.

Reviewer 3: No, it doesn't mean that.  A ref parameter of a struct type can generally be reassigned.  This is a separate constraint, not an impliciation.

Author: Removed that phrase.

- one of the following value types: `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `char`, `float`, `double`, `decimal`, `bool,`; or
- any enumeration type.

### §stack-allocation Stack allocation

> A stack allocation expression allocates a block of memory from the execution stack. The ***execution stack*** is an area of memory where local variables are stored. The execution stack is not part of the managed heap. The memory used for local variable storage is automatically recovered when the current function returns.

Reviewer 2: Not reviewing the whole PR right now, but I like this as a succinct description of the stack.

> The result of a stack allocation may not be copied out of its safe-context (§safe-context-rules). The safe context rules for a stack allocation expression are described in §safe-context-rules-stackalloc.

Reviewer 3: It is an error if you do?

Author: Removed the first sentence. (Yes, it's an error, and the restrictions are described in the xref.)

```ANTLR
stackalloc_expression
    : 'stackalloc' unmanaged_type '[' expression ']'
    | 'stackalloc' unmanaged_type? '[' expression? ']' stackalloc_initializer
    ;

stackalloc_initializer
     : '{' stackalloc_initializer_element_list '}'
     ;

stackalloc_initializer_element_list
     : stackalloc_element_initializer (',' stackalloc_element_initializer)* ','?
     ;
    
stackalloc_element_initializer
    : expression
    ;
```

The *unmanaged_type* ([§8.8](types.md#88-unmanaged-types)) indicates the type of the items that will be stored in the newly allocated location, and the *expression* indicates the number of these items. Taken together, these specify the required allocation size. As the size of a stack allocation cannot be negative, it is a compile-time error to specify the number of items as a *constant_expression* that evaluates to a negative value.

> If *unmanaged_type* is omitted, it is inferred from the corresponding *stackalloc_initializer*. If *expression* is omitted from *stackalloc_expression*, it is inferred to be the number of *stackalloc_element_initializer*s in the corresponding *stackalloc_initializer*.

Reviewer: >it is inferred from the corresponding *stackalloc_initializer* [](http://example.com/codeflow?start=32&length=62)

Should this indicate that "best common type" is used, similar to the description in 12.8.16.5 array creation expressions?

Author: Good catch, but I chose 12.6.3.15 instead as where best common type is defined.

> When a *stackalloc_expression* includes both *expression* and *stackalloc_initializer*, the *expression* shall be a *constant_expression* and the number of elements in that *stackalloc_initializer* shall match the value of *expression*.

Reviewer 3: What if it contains neither?

If `E` is a negative value, then the behavior is undefined. If `E` is zero, then no allocation is made, and the value returned is implementation-defined. If there is not enough memory available to allocate a block of the given size, a `System.StackOverflowException` is thrown.

When *stackalloc_initializer* is present, the *stackalloc_initializer_element_list* shall consist of a sequence of expressions, each having an implicit conversion to *unmanaged_type* ([§10.2](conversions.md#102-implicit-conversions)). The expressions initialize elements in the allocated memory in increasing order, starting with the element at index zero. In the absence of a *stackalloc_initializer*, the content of the newly allocated memory is undefined.

Access via an instance of `System.Span<T>` to the elements of an allocated block is range checked.

Stack allocation initializers are not permitted in `catch` or `finally` blocks ([§13.11](statements.md#1311-the-try-statement)).

> > *Note*: Stack allocation initializers are allowed in `async` methods, but their return value can't be assigned. Neither pointers nor `ref struct` types, like `Span<T>` are allowed in `async` methods. *end note*

Reviewer: >but their return value can't be assigned [](http://example.com/codeflow?start=72&length=40)

What does this mean?

Author: Admittedly not great wording. It's meant to clarify this situation:

```csharp
// This is legal, stackalloc result not assigned to a local variable.
public async Task<int> M4()
{
    _ = stackalloc int[3];
    await Task.Delay(1000);
    return 5;
}

// This is not legal. stackalloc result assigned to local variable, and captured in state machine:
public async Task<int> M4()
{
    Span<int> storage = stackalloc int[3]; // error can't declare a ref struct in an async method
    await Task.Delay(1000);
    return 5;
}

// This is legal. Stackalloc result passed to synchronous method, but not assigned to local variable,
// not captured in state machine:
async Task<int> M4()
{
    int r = M5(stackalloc int[3]);
    await Task.Delay(1000);
    return 5;
}

int M5(Span<int> arg)
{
    return arg.Length;
}
```

Reviewer: Thanks for clarifying. I was confused since I thought "their return value" referred to the return value of the `async` method. Would it make sense to mention the stack allocation explicitly? Perhaps "but the stack allocated result cannot be assigned".

> > *Note*: Stack allocation initializers are allowed in `async` methods, but the stack allocation result can't be assigned. Neither pointers nor `ref struct` types, like `Span<T>` are allowed in `async` methods. *end note*

Reviewer 3: What does "can't be assigned" mean?

Author: would it be better to say "can't be captured"?

Reviewer 2: I'm less keen on captured, as if I'm not using anything that *looks* like it's capturing something, it would be odd to get a complaint about it.

Reviewer 2: @BillWagner: Agreed in the meeting to just delete the note, as it's too hard to describe correctly.

<!-- markdownlint-disable MD028 -->

<!-- markdownlint-enable MD028 -->
> *Note*: There is no way to explicitly free memory allocated using `stackalloc`. *end note*

All stack-allocated memory blocks created during the execution of a function member are automatically discarded when that function member returns.

Except for the `stackalloc` operator, C# provides no predefined constructs for managing non-garbage collected memory. Such services are typically provided by supporting class libraries or imported directly from the underlying operating system.

> *Example*:
>
> ```csharp

> > // memory uninitialized

Reviewer 2: Nit: could we start all of these comments with a capital letter?

> Span<int> span1 = stackalloc int[3];
> // memory initialized
> Span<int> span2 = stackalloc int[3] { -10, -15, -30 };
> // type int is inferred
> Span<int> span3 = stackalloc[] { 11, 12, 13 };
> // error; result is int*, not allowed in a safe context

> > var spn4 = stackalloc[] { 11, 12, 13 };

Reviewer: >spn4 [](http://example.com/codeflow?start=6&length=4)

`span4` ?

Reviewer 4: That would probably be an improvement 😉

If `E` is a negative value, then the behavior is undefined. If `E` is zero, then no allocation is made, and the value returned is implementation-defined. If there is not enough memory available to allocate a block of the given size, a `System.StackOverflowException` is thrown.

When *stackalloc_initializer* is present, the *stackalloc_initializer_element_list* shall consist of a sequence of expressions, each having an implicit conversion to *unmanaged_type* ([§10.2](conversions.md#102-implicit-conversions)). The expressions initialize elements in the allocated memory in increasing order, starting with the element at index zero. In the absence of a *stackalloc_initializer*, the content of the newly allocated memory is undefined.

Access via an instance of `System.Span<T>` to the elements of an allocated block is range checked.

Stack allocation initializers are not permitted in `catch` or `finally` blocks ([§13.11](statements.md#1311-the-try-statement)).

> > *Note*: Stack allocation initializers are allowed in `async` methods, but the stack allocation result can't be assigned. Neither pointers nor `ref struct` types, like `Span<T>` are allowed in `async` methods. *end note*

Author: ```suggestion
> *Note*: Stack allocation initializers are allowed in `async` methods, but the stack allocation result can't be captured. Neither pointers nor `ref struct` types, like `Span<T>` are allowed in `async` methods. *end note*
```

```ANTLR
conditional_expression
    : null_coalescing_expression
    | null_coalescing_expression '?' 'ref'? expression ':' 'ref'? expression
    ;
```

> `ref` shall either be present for both *expression*s or neither. A throw expression (§12.16) is not allowed in a conditional operator if `ref` is present.

Reviewer 4: See general review comments

```suggestion
    | null_coalescing_expression '?' expression ':' expression
    | null_coalescing_expression '?' 'ref'? variable_reference ':' 'ref'?     ;

A throw expression (§12.16) is not allowed in a conditional operator if `ref` is present.
```

```ANTLR
conditional_expression
    : null_coalescing_expression
    | null_coalescing_expression '?' expression ':' expression

>     | null_coalescing_expression '?' 'ref'? variable_reference ':' 'ref'?     ;

Reviewer 4: Typo

```suggestion
    | null_coalescing_expression '?' 'ref' variable_reference ':' 'ref' variable_reference
    ;
```

If `ref` is present:

- Both *expression*s shall be variables and an identity conversion exists between the types, and type of the result can be either type. If either type is `dynamic`, type inference prefers `dynamic` (§8.7).
- The result is a variable, which is writeable if both *expression* variables are writeable.

The run-time processing of a ref conditional expression of the form `b ? ref x : ref y` consists of the following steps:

- First, `b` is evaluated, and the `bool` value of `b` is determined:
  - If an implicit conversion from the type of `b` to `bool` exists, then this implicit conversion is performed to produce a `bool` value.
  - Otherwise, the `operator true` defined by the type of `b` is invoked to produce a `bool` value.
- If the `bool` value produced by the step above is `true`, then `x` is evaluated and the resulting variable reference becomes the result of the conditional expression.
- Otherwise, `y` is evaluated and the resulting variable reference becomes the result of the conditional expression.

> > *Note:* When `ref` is present, the expression is a reference variable, and can be assigned using the `= ref` operator. *end note*

Reviewer 4: The result is a variable reference which can be assigned to a reference variable or passed as a parameter

```suggestion
> *Note:* When `ref` is present, the *conditional_expression* returns a variable reference, which can be assigned to a reference variable using the `= ref` operator or passed as a reference/input/output parameter. *end note*
```

The first operand of the `?:` operator shall be an expression that can be implicitly converted to `bool`, or an expression of a type that implements `operator true`. If neither of these requirements is satisfied, a compile-time error occurs.

If `ref` is present:

> - An identity conversion must exist between the types of the two *variable_reference*s, and type of the result can be either type. If either type is `dynamic`, type inference prefers `dynamic` (§8.7).

Reviewer 3: What does it mean that the type of the result can be either type?  That the compiler gets to decide?  That it is not defined which becomes the type?

Reviewer 4: I’m all for rigor in specification, some might say pedantically so 😉, but I think I’d let this one slide.

Reviewer 2: This is an issue not just for ref, but regular conditionals. For example:

```csharp
bool b = true;
var z = b ? (a: 1, b: 2) : (a: 3, d: 4);
```

The type of `z` is `(a: int, int)`... that needs to be specified somewhere.

Bill has created #803 for this. We need to fix this for C# 7, both for ref and non-ref.

Reviewer 3: See https://github.com/dotnet/roslyn/blob/e006309c2c8748452b7bb404ad07271e8e029d37/src/Compilers/CSharp/Portable/Binder/Semantics/BestTypeInferrer.cs#L251 for the relevant compiler code.  Merging of equivalent types should be specified, but can reasonably be deferred until after v7 ships.

> - Both *expression*s shall be variables and an identity conversion exists between the types, and type of the result can be either type. If either type is `dynamic`, type inference prefers `dynamic` (§8.7).
> - The result is a variable, which is writeable if both *expression* variables are writeable.

Reviewer 4: Need to update to match grammar:

```suggestion
- An identity conversion must exist between the types of the two *variable_reference*s, and type of the result can be either type. If either type is `dynamic`, type inference prefers `dynamic` (§8.7).
- The result is a variable reference, which is writeable if both *variable_reference*s  are writeable.
```

- If the `bool` value produced by the step above is `true`, then `x` is evaluated and the resulting variable reference becomes the result of the conditional expression.
- Otherwise, `y` is evaluated and the resulting variable reference becomes the result of the conditional expression.

> *Note:* When `ref` is present, the *conditional_expression* returns a variable reference, which can be assigned to a reference variable using the `= ref` operator or passed as a reference/input/output parameter. *end note*

> If `ref` is not present, the second and third operands, `x` and `y`, of the `?:` operator control the type of the conditional expression:

Reviewer 4: And if `ref` is present?

Something along the lines of the two *variable_reference*s must be of the same type, *T*, and the type of the expression is *reference to variable of type T* maybe?

Also shouldn't this para, and any added one for the `ref` case, come before the preceding *Note*?

Author: I rearranged the few paragraphs here. The order is now:

1. Type of the result:
    a. For `ref` conditional
    b. note on variable reference
    c. For non-ref conditional
2. Runtime processing
    a. for `ref `conditional
    b. for non-ref conditional.

- If only one of `x` and `y` has a type, and both `x` and `y` are implicitly convertible to that type, then that is the type of the conditional expression.
- Otherwise, no expression type can be determined, and a compile-time error occurs.

The run-time processing of a ref conditional expression of the form `b ? ref x : ref y` consists of the following steps:

- First, `b` is evaluated, and the `bool` value of `b` is determined:
  - If an implicit conversion from the type of `b` to `bool` exists, then this implicit conversion is performed to produce a `bool` value.

>   - Otherwise, the `operator true` defined by the type of `b` is invoked to produce a `bool` value.

Reviewer 3: What if it doesn't exist (there is no such operator)?

Reviewer 4: It’s a compile-time requirement that it does exist if there is no implicit conversion, its up on line 4717 – just further away now `ref` stuff has been added.

Might be worth considering that increased distance, but I’ll leave that decision to the talented wordsmiths among us!

Reviewer 2: No change needed.

anonymous_function_body
    : null_conditional_invocation_expression

>     | 'ref'? expression

Reviewer 4: See general review comments

```suggestion
    | expression
    | 'ref' variable_reference
```

The body (*expression* or *block*) of an anonymous function is subject to the following rules:

- If the anonymous function includes a signature, the parameters specified in the signature are available in the body. If the anonymous function has no signature it can be converted to a delegate type or expression type having parameters ([§10.7](conversions.md#107-anonymous-function-conversions)), but the parameters cannot be accessed in the body.

> - Except for `in`, `out` or `ref` parameters specified in the signature (if any) of the nearest enclosing anonymous function, it is a compile-time error for the body to access an `in`, `out`, or `ref` parameter.

Reviewer 3: or a variable of a ref struct type?

Author: verified and updated. Added an additional bullet.

### 12.21.1 General

> All but one of the assignment operators assigns a new value to a variable, a property, an event, or an indexer element. The exception, `= ref`, creates an alias to a variable.

Reviewer 4: Do we use “alias” like this elsewhere?

```suggestion
All but one of the assignment operators assigns a new value to a variable, a property, an event, or an indexer element. The exception, `= ref`, assigns a variable reference (§9.5) to a reference variable (§9.7?).
```

```ANTLR
assignment
    : unary_expression assignment_operator expression
    ;

assignment_operator
    : '=' 'ref'? | '+=' | '-=' | '*=' | '/=' | '%=' | '&=' | '|=' | '^=' | '<<='
    | right_shift_assignment
    ;
```

The left operand of an assignment shall be an expression classified as a variable, or, except for `= ref`, a property access, an indexer access, an event access or a tuple. A declaration expression is not directly permitted as a left operand, but may occur as a step in the evaluation of a deconstructing assignment.

The `=` operator is called the ***simple assignment operator***. It assigns the value or values of the right operand to the variable, property, indexer element or tuple elements given by the left operand. The left operand of the simple assignment operator shall not be an event access (except as described in [§15.8.2](classes.md#1582-field-like-events)). The simple assignment operator is described in [§12.21.2](expressions.md#12212-simple-assignment).

> The operator `= ref`  is called the ***ref assignment operator***. It makes the variable designated by the left operand, an alias for the variable designated by the right operand. The ref assignment operator is described in §ref-assignment-new-clause.

Reviewer 4: See line 6208 above and "Ref assignment" below.

```suggestion
The operator `= ref`  is called the ***ref assignment operator***. It makes the right operand, which must be a *variable_reference* (§9.5), the referent of the reference variable designated by the left operand. The ref assignment operator is described in §ref-assignment-new-clause.
```

A x-ref could be added after "reference variable"

>
> *end example*

### §ref-assignment-new-clause Ref assignment

The `= ref` operator is known as the *ref assignment* operator.

> The left operand shall be an expression that binds to a ref local variable, a ref parameter (other than `this`), or an out parameter. The right operand shall be an expression that yields a *variable_reference* designating a value of the same type as the left operand.

Reviewer 4: Adding input parameters and fixing up terminology

```suggestion
The left operand shall be an expression that binds to a reference variable (§fix_me_up_probably_9.7), a reference parameter (other than `this`), an output parameter, or and input parameter. The right operand shall be an expression that yields a *variable_reference* designating a value of the same type as the left operand.
```

The operation makes the left operand an alias of the right operand variable. The alias may be made read-only even if the right operand variable is writeable.

The ref assignment operator yields a *variable_reference* of the assigned type. It is writeable if the left operand is writeable.

The ref assignment operator shall not read the storage location referenced by the right hand operand.

> *Example*: Here are some examples of using `= ref`:
>
> ```csharp
> public static int M1() { … }
> public static ref int M2() { … }
> public static ref uint M2u() { … }
> public static ref readonly int M3() { … }
> 
> int v = 42;
> ref int r1 = ref v; // OK, r1 refers to v, which has value 42
> r1 = ref M1();      // Error; M1 returns a value, not a reference
> r1 = ref M2();      // OK; makes an alias
> r1 = ref M2u();     // Error; lhs and rhs have different types
> r1 = ref M3();    // error; M3 returns a ref readonly, which r1 cannot honor

> > ref readonly r2 = …;

Reviewer: >r2 = …; [](http://example.com/codeflow?start=15&length=7)

The type is missing. Perhaps also add an explicit RHS rather than "...". For instance: `ref readonly int r2 = ref v;`

>
> *end example*

### §ref-assignment-new-clause Ref assignment

The `= ref` operator is known as the *ref assignment* operator.

> The left operand shall be an expression that binds to a reference variable (§ref-span-safety), a reference parameter (other than `this`), an output parameter, or and input parameter. The right operand shall be an expression that yields a *variable_reference* designating a value of the same type as the left operand.

Reviewer 3: Typo "or and input parameter"

> It is a compile time error if the ref-safe-context of the left operand is wider than the ref-safe-context of the left operand (§ref-safe-contexts).

Reviewer 3: Should one of these lefts be right?  And the other wrong?

Reviewer 3: There should be a cross-reference to the definition of ref-safe-context.

Reviewer 4: Well the second left should be right, that's a typo, but I'm lost on the other one being wrong! 😉

There is a cross-reference, are you suggesting it be moved to be after “ref-safe-context” itself and not placed at the end of the sentence? Makes sense to me.

```suggestion
It is a compile time error if the ref-safe-context (`§ref-safe-contexts`) of the left operand is wider than the ref-safe-context of the right operand.
```

The right operand shall be definitely assigned at the point of the ref assignment.

When the left operand binds to an `out` parameter, it is an error if that `out` parameter has not been definitely assigned at the beginning of the ref assignment operator.

If the left operand is a writeable ref (i.e., it designates anything other than a `ref readonly` local or  `in` parameter), then the right operand shall be a writeable *variable_reference*. If the right operand variable is writeable, the left operand may be declared `ref` or `ref readonly`.

The operation makes the left operand an alias of the right operand variable. The alias may be made read-only even if the right operand variable is writeable.

The ref assignment operator yields a *variable_reference* of the assigned type. It is writeable if the left operand is writeable.

> The ref assignment operator shall not read the storage location referenced by the right hand operand.

Reviewer 3: Who is this restricting?  If the compiler-writer, you should instead say that it "does not".

Reviewer 4: Or “must not”?

> *Example*: Here are some examples of using `= ref`:
>
> ```csharp
> public static int M1() { … }
> public static ref int M2() { … }
> public static ref uint M2u() { … }
> public static ref readonly int M3() { … }
> 
> int v = 42;
> ref int r1 = ref v; // OK, r1 refers to v, which has value 42
> r1 = ref M1();      // Error; M1 returns a value, not a reference
> r1 = ref M2();      // OK; makes an alias
> r1 = ref M2u();     // Error; lhs and rhs have different types
> r1 = ref M3();    // error; M3 returns a ref readonly, which r1 cannot honor
> ref readonly r2 = ref v; // OK; make readonly alias to ref
> r2 = ref M2();      // OK; makes an alias, adding read-only protection
> r2 = ref M3();      // OK; makes an alias and honors the read-only
> r2 = ref (r1 = ref M2());  // OK; r1 is an alias to a writable variable,
>               // r2 is an alias (with read-only access) to the same variable
> ```
>
> *end example*

> 

Reviewer 2: I had an action item to add a note here. I don't think we *have* to do before we ship C# 7, but it *could* be something like:

```text
*Note*: When reading code using an `= ref` operator, it can be tempting to read
the `ref` part as being part of the operand. This is particularly confusing when
the operand is a conditional `?:` expression. For example, when reading
`ref int a = ref b ? ref x : ref y;` it's important to read this as `= ref` being the operator,
and `b ? ref x : ref y` being the right operand. Importantly, the expression
`ref b` is *not* part of that statement, even though it might appear so at first glance.
```

I'm not convinced that's good enough to include yet, but I'd welcome feedback about whether it's close enough to be included with a bit of mob-wordsmithing.

Author: I like it. I think it's clarified with the addition of parentheses in the example, after explaining how it's parsed:

```csharp
ref int a = ref (b ? ref x : ref y);
```

Reviewer 2: Yes! That definitely helps a lot.

Reviewer 2: Suggestion committed.

The ref assignment operator yields a *variable_reference* of the assigned type. It is writeable if the left operand is writeable.

The ref assignment operator must not read the storage location referenced by the right hand operand.

> *Example*: Here are some examples of using `= ref`:
>
> ```csharp
> public static int M1() { … }
> public static ref int M2() { … }
> public static ref uint M2u() { … }
> public static ref readonly int M3() { … }
> 
> int v = 42;
> ref int r1 = ref v; // OK, r1 refers to v, which has value 42
> r1 = ref M1();      // Error; M1 returns a value, not a reference
> r1 = ref M2();      // OK; makes an alias
> r1 = ref M2u();     // Error; lhs and rhs have different types
> r1 = ref M3();    // error; M3 returns a ref readonly, which r1 cannot honor
> ref readonly r2 = ref v; // OK; make readonly alias to ref
> r2 = ref M2();      // OK; makes an alias, adding read-only protection
> r2 = ref M3();      // OK; makes an alias and honors the read-only
> r2 = ref (r1 = ref M2());  // OK; r1 is an alias to a writable variable,
>               // r2 is an alias (with read-only access) to the same variable
> ```
>
> *end example*

> 

Author: Adding Jon's text as a suggestion:

```suggestion
<!-- markdownlint-disable MD028 -->

<!-- markdownlint-enable MD028 -->
> *Note*: When reading code using an `= ref` operator, it can be tempting to read the `ref` part as being part of the operand. This is particularly confusing when the operand is a conditional `?:` expression. For example, when reading `ref int a = ref b ? ref x : ref y;` it's important to read this as `= ref` being the operator, and `b ? ref x : ref y` being the right operand: `ref int a = ref (b ? ref x : ref y`);`. Importantly, the expression `ref b` is *not* part of that statement, even though it might appear so at first glance. *end note*
```
